### PR TITLE
Support sending telemetry event reports

### PR DESCRIPTION
**Pull request by mkindahl**
Add table `_timescaledb_catalog.telemetry_event` table containing
events that should be sent out with telemetry reports. The table will
be truncated after report being generated.

### CODE REVIEW

File: src/telemetry/telemetry_metadata.c [hunk 1]
```diff
@@ -15,6 +17,72 @@
...
+		.relations = list_make1(&rv),
+		.behavior = DROP_RESTRICT,
+	});
+}
+
+void
+ts_telemetry_events_add(JsonbParseState *state)
+{
+	ScanIterator iterator =
+		ts_scan_iterator_create(TELEMETRY_EVENT, AccessShareLock, CurrentMemoryContext);
```
Comment (kgyrtkirk) [inline src/telemetry/telemetry_metadata.c:49]: there doesn't seem to be any limit on the amount of these events being added - shouldn't there be one?

I wonder how realistic of the case of having a request pumped up by many events; failing to be processed on the other end - so that it will just keep failing again and again...

File: sql/updates/latest-dev.sql [hunk 2]
```diff
@@ -55,3 +55,21 @@ ALTER FUNCTION _timescaledb_internal.rxid_out(@extschema@.rxid) SET SCHEMA _time
...
+       body jsonb NOT NULL
+);
+
+SELECT pg_catalog.pg_extension_config_dump('_timescaledb_catalog.telemetry_event', '');
+
+GRANT SELECT ON _timescaledb_catalog.telemetry_event TO PUBLIC;
+
+DROP FUNCTION IF EXISTS @extschema@.get_telemetry_report();
+
+CREATE FUNCTION @extschema@.get_telemetry_report(truncate_telemetry_events bool = false)
```
Comment (kgyrtkirk) [inline sql/updates/latest-dev.sql:73]: is "wiping clean a table" is something which should happen in a `get` function?

Comment (kgyrtkirk) [reply sql/updates/latest-dev.sql:73]: This function declaration seem to be part of the public api - If its used only in testing; it should be declared only during that...we are also maintaining it via drop/creates between versions - if thats not warranted ; why keep it?

I wonder if there are any obstacles which prevents the tests do something like:
```
truncate _timescaledb_catalog.telemetry_event;
```

Comment (mkindahl) [reply sql/updates/latest-dev.sql:73]: > This function declaration seem to be part of the public api - If its used only in testing; it should be declared only during that...we are also maintaining it via drop/creates between versions - if thats not warranted ; why keep it?

Hmm... good point.

File: src/telemetry/telemetry_metadata.c (hunk 1 above)
Comment (kgyrtkirk) [reply src/telemetry/telemetry_metadata.c:49]: I'm just saying that it doesn't have any bounds - which might cause it to hit back in possibly odd situations....

let me rephrase it differently: if for some reason the target service where this is connecting can't be reached (say firewall rules) - the telemetry events will keep accumlating ; and the system will build a bigger and bigger request (for no good reason).

I don't know how frequently events will be produced and at what rate...
if you think its not worth the effort - it could be left as-is.

File: sql/updates/latest-dev.sql (hunk 2 above)
Comment (mkindahl) [reply sql/updates/latest-dev.sql]: Removed the changes to the `get_telemetry_report` function.

File: src/telemetry/telemetry_metadata.c (hunk 1 above)
Comment (mkindahl) [reply src/telemetry/telemetry_metadata.c:49]: > I don't know how frequently events will be produced and at what rate...
if you think its not worth the effort - it could be left as-is.

It is not yet used by anything and we do not expect this to generate a lot of data.

**Review by kgyrtkirk — APPROVED** [review]

**Review by akuzm — APPROVED** [review]

File: src/telemetry/telemetry_metadata.c [hunk 3]
```diff
@@ -15,6 +17,72 @@
...
+JsonbToJsonbValue(Jsonb *jsonb, JsonbValue *val)
+{
+	val->type = jbvBinary;
+	val->val.binary.data = &jsonb->root;
+	val->val.binary.len = VARSIZE(jsonb) - VARHDRSZ;
+}
+#endif
+
+void
+ts_telemetry_event_truncate(void)
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply src/telemetry/telemetry_metadata.c:32]: There are no good ways to test this AFAICT, but I will see if I can at least get some partial coverage, e.g., make sure that it is truncated after successful sending.

[the remark this answers is not in the document]
Comment (mkindahl) [reply src/telemetry/telemetry_metadata.c:32]: This is only used for the telemetry background worker, so it is hard to add test coverage for it. The existing telemetry tests only create a fake backend to check for errors.

[the remark this answers is not in the document]
Comment (mkindahl) [reply src/telemetry/telemetry_metadata.c:32]: There seems to be no good way to trigger this behavior, but I found a bug in the process. This is fixed now, but I will merge this without a test. I have tested it manually to make sure that it works.

```
mats=# INSERT INTO _timescaledb_catalog.telemetry_event(tag, body) VALUES
    ('test-1', '{"title": "First test value"}'),
    ('test-2', '{"title": "Second test value"}');
INSERT 0 2
mats=# SELECT * FROM _timescaledb_catalog.telemetry_event;
            created            |  tag   |              body              
-------------------------------+--------+--------------------------------
 2023-05-12 15:45:58.611144+02 | test-1 | {"title": "First test value"}
 2023-05-12 15:45:58.611144+02 | test-2 | {"title": "Second test value"}
(2 rows)

mats=# SELECT id AS job_id FROM _timescaledb_config.bgw_job
 WHERE proc_schema = '_timescaledb_internal'
   AND proc_name = 'policy_telemetry' \gset
mats=# CALL run_job(:job_id);
NOTICE:  the "timescaledb" extension is up-to-date
CALL
mats=# SELECT * FROM _timescaledb_catalog.telemetry_event;
 created | tag | body 
---------+-----+------
(0 rows)
```
